pulse-rs/src/stream.rs

```diff
@@ -0,0 +1,413 @@
+        {
+            let mut stm = stream::from_raw_ptr(s);
+            let result = mem::transmute::<_, &F>(&())(&mut stm, nbytes, userdata);
+            mem::forget(stm);
+            result
+        }
+
+        unsafe {
+            ffi::pa_stream_set_read_callback(self.raw_mut(), Some(wrapped::<CB>), userdata);
+        }
+    }
+
+    pub fn cork<CB>(&self, b: i32, _: CB, userdata: *mut c_void) -> Result<Operation>
+        where CB: Fn(&Stream, i32, *mut c_void)
+    {
+        debug_assert_eq!(mem::size_of::<CB>(), 0);
+
+        unsafe extern "C" fn wrapped<F>(s: *mut ffi::pa_stream,
+                                        success: c_int,
+                                        userdata: *mut c_void)
+            where F: Fn(&Stream, i32, *mut c_void)
+        {
+            let mut stm = stream::from_raw_ptr(s);
+            let result = mem::transmute::<_, &F>(&())(&mut stm, success, userdata);
+            mem::forget(stm);
+            result
+        }
+
+        let r = unsafe { ffi::pa_stream_cork(self.raw_mut(), b, Some(wrapped::<CB>), userdata) };
+        if r.is_null() {
+            let err = if let Some(c) = self.get_context() {
+                c.errno()
+            } else {
+                ffi::PA_ERR_UNKNOWN
+            };
+            return Err(ErrorCode::from_error_code(err));
+        }
+        Ok(unsafe { operation::from_raw_ptr(r) })
+    }
+
+    pub fn get_time(&self) -> Result<(u64)> {
+        let mut usec: u64 = 0;
+        let r = unsafe { ffi::pa_stream_get_time(self.raw_mut(), &mut usec) };
+        error_result!(usec, r)
+    }
+
+    pub fn get_latency(&self) -> Result<(u64, bool)> {
+        let mut usec: u64 = 0;
+        let mut negative: i32 = 0;
+        let r = unsafe { ffi::pa_stream_get_latency(self.raw_mut(), &mut usec, &mut negative) };
+        error_result!((usec, negative != 0), r)
+    }
+
+    pub fn get_sample_spec(&self) -> &SampleSpec {
+        unsafe {
+            let ptr = ffi::pa_stream_get_sample_spec(self.raw_mut());
+            debug_assert!(!ptr.is_null());
+            &*ptr
+        }
+    }
+
+    pub fn get_channel_map(&self) -> &ChannelMap {
+        unsafe {
+            let ptr = ffi::pa_stream_get_channel_map(self.raw_mut());
+            debug_assert!(!ptr.is_null());
+            &*ptr
+        }
+    }
+
+    pub fn get_buffer_attr(&self) -> &BufferAttr {
+        unsafe {
+            let ptr = ffi::pa_stream_get_buffer_attr(self.raw_mut());
+            debug_assert!(!ptr.is_null());
+            &*ptr
+        }
+    }
+}
+
+impl Clone for Stream {
+    fn clone(&self) -> Self {
+        unsafe { stream::from_raw_ptr(ffi::pa_stream_ref(self.raw_mut())) }
+    }
+}
+
+impl ::std::default::Default for Stream {
+    fn default() -> Self {
+        Stream(::std::ptr::null_mut())
+    }
+}
+
+impl Drop for Stream {
+    fn drop(&mut self) {
+        if !self.is_null() {
+            unsafe {
+                ffi::pa_stream_unref(self.raw_mut());
```

Comment:
Same question as for the Context regarding the drop vs shutdown/destroy tracking.

Is there value in implementing clone for these things? I don't see any current use of `.clone()`.